### PR TITLE
Add dlarts

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
Not ready for submission. 

Dlarts is necessary for Dgecon (Compute condition number from QR facotrization) and Dtrcon (compute condition of triangular matrix, which can be used from LU to compute the condition number). 

In the normal case, this function mostly just calls Dtrsv. In the case where the matrix is badly conditioned, it does all sorts of crazy things. Any ideas on testing? For the badly conditioned cases in BLAS I created the ill conditioned cases by hand (for the Givens rotation cases). Here though, there are many possible code paths, so I can't construct a suffient set by hand. Additionally, it's meant to protect against bad scaling, so we can't compare the output to Dtrsv (because it shouldn't be the same), and I'm worried about testing that A * x = b, because presumably we would need also need to be careful with scaling. At the least, I'll need to do a second pass with visual inspection.

**Comment by kortschak** [comment]
You should be able to use dvykov's go-fuzz to build a minimal corpus that tests all paths with some fiddling.

**Comment by jonlawlor** [comment]
If you're going to try out fuzz, I think you'll first want to write byte encoders / decoders for the various blas types.

**Comment by kortschak** [comment]
We have one for Dense already, the rest can be copied from a created Dense.

**Comment by btracey** [comment]
> We have one

One what? A go-fuzz example, or a comparison for triangular solve?

**Comment by kortschak** [comment]
We have a binary encoder. This means that a fuzzer can be written with a corpus of encoded matrices from the original handwritten test cases. The other matrix types can be created from the decoded Dense.

**Comment by btracey** [comment]
Okay, I ran a fuzzer against it for a while so I think it's free of panics. Unfortunately, there is no cgo version of the algorithm so I cannot compare outputs.

PTAL

**Comment by kortschak** [comment]
Pity Dlatrs doesn't exist in the cgo. I'll have a look some time over the weekend.

**Comment by kortschak** [comment]
Change the PR title to reflect the function name?

File: native/dlatrs.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+		panic(badUplo)
+	}
+	if trans != blas.Trans && trans != blas.NoTrans {
+		panic(badTrans)
+	}
+	if diag != blas.Unit && diag != blas.NonUnit {
+		panic(badDiag)
+	}
+	upper := uplo == blas.Upper
+	notrans := trans == blas.NoTrans
```
Comment (kortschak) [inline native/dlatrs.go]: s/notrans/noTrans/g
s/nonunit/nonUnit/g

File: native/dlatrs.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+	nonunit := diag == blas.NonUnit
+
+	if n < 0 {
+		panic(nLT0)
+	}
+	checkMatrix(n, n, a, lda)
+	// Why incX isn't specified like every other blas/lapcak function I have no idea.
```
Comment (kortschak) [inline native/dlatrs.go]: s/lapcak/lapack/

Diatribe has merit, but could be more informative - make the concern clearer.

File: native/dlatrs.go [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+	checkVector(n, cnorm, 1)
+
+	if n == 0 {
+		return
+	}
+	scale = 1
+	bi := blas64.Implementation()
+	if !normin {
+		if upper {
+			for j := 0; j < n; j++ {
```
Comment (kortschak) [inline native/dlatrs.go]: I take it the lda here is because we are row major.

File: native/dlatrs.go [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+			jinc = -1
+		} else {
+			jfirst = 0
+			jlast = n - 1
+			jinc = 1
+		}
+		// Compute the growth in A * x = b.
+		if tscal != 1 {
+			grow = 0
+			goto Out
```
Comment (kortschak) [inline native/dlatrs.go]: s/Out/finish/g or some near equivalent.

File: native/dlatrs.go [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+			for j := jfirst; j != jlast; j += jinc {
+				if grow <= smlnum {
+					goto Out
+				}
+				xj := 1 + cnorm[j]
+				grow /= xj
+			}
+		}
+	}
+Out:
```
Comment (kortschak) [inline native/dlatrs.go]: Newline before this.

File: native/dlatrs.go [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+						bi.Dscal(n, rec, x, 1)
+						scale *= rec
+					}
+				} else if xj*cnorm[j] > bignum-xmax {
+					bi.Dscal(n, 0.5, x, 1)
+					scale *= 0.5
+				}
+				if upper {
+					if j > 0 {
+						bi.Daxpy(j, -x[j]*tscal, a[j:], lda, x, 1)
```
Comment (kortschak) [inline native/dlatrs.go]: Should this be a[:j]?

**Comment by kortschak** [comment]
LGTM with minor comments.

File: native/dlatrs.go (hunk 1 above)
Comment (btracey) [inline native/dlatrs.go]: Done.

File: native/dlatrs.go (hunk 2 above)
Comment (btracey) [inline native/dlatrs.go]: Just removed it.

File: native/dlatrs.go (hunk 3 above)
Comment (btracey) [inline native/dlatrs.go]: Yep. It's computing the sum over the column.

File: native/dlatrs.go (hunk 5 above)
Comment (btracey) [inline native/dlatrs.go]: Done.

File: native/dlatrs.go (hunk 4 above)
Comment (btracey) [inline native/dlatrs.go]: Done.

**Comment by kortschak** [comment]
s/dlarts/dlatrs/ in name, but doesn't matter.

**Pull request merged**